core/unix/signal_linux_aarch64.c

```diff
@@ -282,40 +282,109 @@ mcontext_to_sigcontext_simd(sig_full_cxt_t *sc_full, priv_mcontext_t *mc)
 {
 #ifdef DR_HOST_NOT_TARGET
     ASSERT_NOT_REACHED();
-#endif
+#else
+#    define GET_RECORD_AT_OFFSET(byte_offset) \
+        (void *)((byte *)sc_full->fp_simd_state + byte_offset)
+
     /* sig_full_initialize() will have set the fp_simd_state pointer in the
      * user level machine context's (uc_mcontext) to __reserved.
      */
-    struct fpsimd_context *fpc = (struct fpsimd_context *)sc_full->fp_simd_state;
-    if (fpc == NULL)
+    size_t offset = 0;
+    struct _aarch64_ctx *next_head = GET_RECORD_AT_OFFSET(offset);
+
+    if (next_head == NULL) {
         return;
-    fpc->head.magic = FPSIMD_MAGIC;
-    fpc->head.size = sizeof(struct fpsimd_context);
-    fpc->fpsr = mc->fpsr;
-    fpc->fpcr = mc->fpcr;
-    ASSERT(sizeof(fpc->vregs) == (sizeof(mc->simd->q) * MCXT_NUM_SIMD_SVE_SLOTS));
-    int i;
-    for (i = 0; i < MCXT_NUM_SIMD_SVE_SLOTS; i++) {
-        memcpy(&fpc->vregs[i], &mc->simd[i].u32[0], sizeof(fpc->vregs[i]));
     }
 
-#ifndef DR_HOST_NOT_TARGET
-    if (proc_has_feature(FEATURE_SVE)) {
-        struct _aarch64_ctx *esr = (void *)((byte *)fpc + sizeof(struct fpsimd_context));
-        esr->magic = ESR_MAGIC;
-        esr->size = sizeof(struct esr_context);
-
-        struct sve_context *sve = (void *)((byte *)esr + sizeof(struct esr_context));
-        // Set other fields of sve_context to zero. New fields may be added and
-        // unexpected values in those fields may cause problems. This is a small
-        // struct so the compiler will allocate it to a virtual register and
-        // optimise this initialisation.
+    struct fpsimd_context *fpsimd = NULL;
+    struct sve_context *sve = NULL;
+
+    const struct _aarch64_ctx null_ctx = { 0 };
+
+    /* Search for the context records we are interested in. */
+    while (next_head->magic != 0) {
+        switch (next_head->magic) {
+        case FPSIMD_MAGIC: fpsimd = (void *)next_head; break;
+        case SVE_MAGIC: sve = (void *)next_head; break;
+        default: break;
+        }
+        offset += next_head->size;
+        next_head = GET_RECORD_AT_OFFSET(offset);
+    }
+
+    if (sve != NULL) {
+        /* Check that we have a full SVE record with space for the register data. */
+        ASSERT(sve->vl == proc_get_vector_length_bytes());
+        if (sve->head.size < SVE_SIG_CONTEXT_SIZE(sve_vecquad_from_veclen(sve->vl))) {
+            /* sigcontext has an sve_context record but it doesn't contain any register
+             * data. We need to remove it and add a new larger record with space for the
+             * register data.
+             */
+            size_t sve_record_size = sve->head.size;
+
+            byte *dst = (void *)sve;
+            byte *src = dst + sve_record_size;
+            size_t length =
+                (ptr_uint_t)next_head + sizeof(struct _aarch64_ctx) - (ptr_uint_t)src;
+            /* We can't use memcpy because the ranges overlap. */
```

Comment:
Use memmove. We have an internal copy. It is not very optimized right now but better to go through a central routine so if we optimize later all uses benefit.